core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```diff
@@ -2,59 +2,66 @@ package org.broadinstitute.dsde.rawls.snapshot
-      stubExists <- workspaceStubExists(workspaceId, userInfo)
-      _ <- if(!stubExists) { workspaceManagerDAO.createWorkspace(workspaceId, userInfo) } else Future.successful()
-      dataRepoReference = JsObject.apply(("instance", JsString(terraDataRepoUrl)), ("snapshot", JsString(snapshot.snapshotId)))
-      res <- workspaceManagerDAO.createDataReference(workspaceId, snapshot.name, DataReferenceType.DataRepoSnapshot.toString, dataRepoReference, CloningInstructions.COPY_NOTHING.toString, userInfo)
-    } yield {
-      DataRepoSnapshotReference(res.getReferenceId.toString, res.getName, res.getWorkspaceId.toString, Option(res.getReferenceType.toString), Option(res.getReference), res.getCloningInstructions.toString)
+    getWorkspaceContextAndPermissions(workspaceName, SamWorkspaceActions.write, Some(WorkspaceAttributeSpecs(all = false))).flatMap { workspaceContext =>
+      if(!workspaceStubExists(workspaceContext.workspaceId, userInfo)) {
+        workspaceManagerDAO.createWorkspace(workspaceContext.workspaceId, getServiceAccountAccessToken, userInfo.accessToken)
+      }
+
+      val dataRepoReference = JsObject.apply(("instance", JsString(terraDataRepoUrl)), ("snapshot", JsString(snapshot.snapshotId)))
+      val ref = workspaceManagerDAO.createDataReference(workspaceContext.workspaceId, snapshot.name, ReferenceTypeEnum.DATAREPOSNAPSHOT.getValue, dataRepoReference, CloningInstructionsEnum.NOTHING.getValue, userInfo.accessToken)
+
+      Future.successful(DataRepoSnapshotReference(ref.getReferenceId.toString, ref.getName, ref.getWorkspaceId.toString, Option(ref.getReferenceType.toString), Option(ref.getReference), ref.getCloningInstructions.toString))
     }
   }
 
   def getSnapshot(workspaceName: WorkspaceName, snapshotId: String): Future[DataRepoSnapshotReference] = {
-    for {
-      workspaceIdOpt <- dataSource.inTransaction { dataAccess => dataAccess.workspaceQuery.getWorkspaceId(workspaceName) }
-      workspaceId = workspaceIdOpt.getOrElse(throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"Workspace $workspaceName not found")))
-      res <- workspaceManagerDAO.getDataReference(workspaceId, UUID.fromString(snapshotId), userInfo)
-    } yield {
-      DataRepoSnapshotReference(res.getReferenceId.toString, res.getName, res.getWorkspaceId.toString, Option(res.getReferenceType.toString), Option(res.getReference), res.getCloningInstructions.toString)
+    getWorkspaceContextAndPermissions(workspaceName, SamWorkspaceActions.read, Some(WorkspaceAttributeSpecs(all = false))).flatMap { workspaceContext =>
+      val ref = workspaceManagerDAO.getDataReference(workspaceContext.workspaceId, UUID.fromString(snapshotId), userInfo.accessToken)
+      Future.successful(DataRepoSnapshotReference(ref.getReferenceId.toString, ref.getName, ref.getWorkspaceId.toString, Option(ref.getReferenceType.toString), Option(ref.getReference), ref.getCloningInstructions.toString))
```

Comment:
if `DataRepoSnapshotReference` had an apply method:
```
def apply(ref: DataReferenceDescription): DataRepoSnapshotReference = {
    new DataRepoSnapshotReference(ref.getReferenceId.toString, ref.getName, ref.getWorkspaceId.toString,
      Option(ref.getReferenceType.toString), Option(ref.getReference), ref.getCloningInstructions.toString)
  }
```
then here and in the return clause for createSnapshot, you could have just:
```
Future.successful(DataRepoSnapshotReference(ref))
```